core/src/util_normalize.cpp

```diff
@@ -10,11 +10,24 @@
 #include "core_icu.h"
 #include "kmx/kmx_xstring.h"
 
+#ifdef __EMSCRIPTEN__
+#include <emscripten.h>
+#include "utfcodec.hpp"
+// JS implementations
+
+EM_JS(char*, NormalizeNFD, (const char* input), {
+  if (!input) return input; // pass through null
+  const instr = Module.UTF8ToString(input);
+  const nfd = instr.normalize("NFD");
```

Comment:
I'm not sure what is happening here. Is ln 21 equivalent to `icu::Normalizer2::getNFDInstance(status);`
where is the ln 21 normalised defined?

Reply:
Yes it's the JS idiom for the same thing.  Ends up calling icu..

Reply:
So... this line is doing the JS call?  I don't see a definition for `instr`, and I'd like to learn more about whatever interface there is that would allow WASM to make native-JS calls.

Reply:
The entire EM_JS block is Javascript.

> I don't see a definition for `instr`

It's on line 20.

> I'd like to learn more about whatever interface there is that would allow WASM to make native-JS calls.

https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html#interacting-with-code-call-javascript-from-native

Reply:
On Emscripten, beware, it's a deep rabbit hole.

Reply:
> > I don't see a definition for `instr`
> 
> It's on line 20.

Kinda passes the buck off to `Module`, though.  It's at least referenced through the link you provided - thanks!  Kinda odd how it just assumes `Module` exists and doesn't talk about it, though.

In case I return to this comment in the future, or someone else wants a reference on it... https://emscripten.org/docs/api_reference/module.html#other-methods

I'm still not clear on where exactly `UTF8ToString` comes from, but I'm willing to handwave that tidbit away for now.

Reply:
Yeah, it's all part of the Emscripten tooling. It's a little magical when you first encounter it.

Reply:
Module is a JavaScript global at that point.

Yeah it's mostly glue and one call. But it will eventually save us a lot in linked libraries.